File: leetcode/lc_038.c
//
// LeetCode
// 38. Count and Say
// https://leetcode.com/problems/count-and-say/
//


char* countAndSay(int n) {
    int i, j, l, count;
    char num, *s, *result;
    
    if (n == 1)
        return "1";
    
    s = countAndSay(n - 1);
    l = strlen(s);
    result = (char *) malloc((2*l + 1) * sizeof(char));
    num = s[0];
    count = 1;
    j = 0;
    
    for (i = 1; i < l; i++) {
        if (s[i] == num) {
            count++;
        } else {
            result[j++] = '0' + count;
            result[j++] = num;

            count = 1;
            num = s[i];
        }
    }
    
    result[j++] = '0' + count;
    result[j++] = num;
    result[j] = '\0';
    
    return result;
}

File: leetcode-swift/lc_038.swift
//
// LeetCode
// 38. Count and Say
// https://leetcode.com/problems/count-and-say/
//


class Solution {
    func countAndSay(_ n: Int) -> String {
        precondition(n > 0, "Input must be a positive interger.")
        
        if n == 1 {
            return "1"
        }
        
        let seq = countAndSay(n - 1)
        var currentIndex = seq.startIndex
        var num = seq[currentIndex]
        var result = ""
        var count = 1
        
        for i in 1..<seq.count {
            let nextIndex = seq.index(seq.startIndex, offsetBy: i)
            
            if seq[nextIndex] == num {
                count += 1
            } else {
                result += "\(count)\(seq[currentIndex])"
                currentIndex = nextIndex
                num = seq[currentIndex]
                count = 1
            }
        }
        
        result += "\(count)\(seq[currentIndex])"
        
        return result
    }
}
